src/api/eth_filter.rs

```diff
@@ -114,104 +93,94 @@ impl<T: Transport, R> BaseFilter<T, R> {
   }
 }
 
-impl<T: Transport, R> Drop for BaseFilter<T, R> {
-  fn drop(&mut self) {
-    let _ = self.uninstall_internal().wait();
-  }
-}
-
-/// Logs Filter
-pub struct LogsFilter<T: Transport> {
-  base: BaseFilter<T, Log>,
-}
-
-impl<T: Transport> FilterHandle<T> for LogsFilter<T> {
-  fn new(id: U256, transport: T) -> Self {
-    LogsFilter {
-      base: BaseFilter {
-        id: id,
-        transport: transport,
-        _output: PhantomData,
-      }
-    }
-  }
-}
-
-impl<T: Transport> LogsFilter<T> {
-  /// Get all logs matching the filter.
+impl<T: Transport> BaseFilter<T, LogsFilter> {
   pub fn logs(&self) -> CallResult<Vec<Log>, T::Out> {
     let id = helpers::serialize(&self.id);
     CallResult::new(self.transport.execute("eth_getFilterLogs", vec![id]))
   }
 }
 
-impl<T: Transport> Deref for LogsFilter<T> {
-  type Target = BaseFilter<T, Log>;
-
-  fn deref(&self) -> &Self::Target {
-    &self.base
+impl<T: Transport, F: FilterInterface> Drop for BaseFilter<T, F> {
+  fn drop(&mut self) {
+    let _ = self.uninstall_internal().wait();
   }
 }
 
-/// New blocks hashes filter.
-pub struct BlocksFilter<T: Transport> {
-  base: BaseFilter<T, H256>,
+fn create_filter<T: Transport + Clone, F: FilterInterface>(t: T, arg: Vec<rpc::Value>) -> CreateFilter<T, F> {
+  CreateFilter {
+    transport: t.clone(),
+    interface: PhantomData,
+    future: CallResult::new(t.execute(F::constructor(), arg))
+  }
 }
 
-impl<T: Transport> FilterHandle<T> for BlocksFilter<T> {
-  fn new(id: U256, transport: T) -> Self {
-    BlocksFilter {
-      base: BaseFilter {
-        id: id,
-        transport: transport,
-        _output: PhantomData,
-      }
-    }
-  }
+pub struct CreateFilter<T: Transport, F: FilterInterface> {
+  transport: T,
+  interface: PhantomData<F>,
+  future: CallResult<U256, T::Out>,
 }
 
-impl<T: Transport> Deref for BlocksFilter<T> {
-  type Target = BaseFilter<T, H256>;
+impl<T, F> Future for CreateFilter<T, F> where
+  T: Transport + Clone,
+  F: FilterInterface
+{
+  type Item = BaseFilter<T, F>;
+  type Error = Error;
 
-  fn deref(&self) -> &Self::Target {
-    &self.base
+  fn poll(&mut self) -> Poll<Self::Item, Error> {
+    match self.future.poll() {
+      Ok(Async::Ready(x)) => Ok(Async::Ready(BaseFilter {
+        id: x,
+        transport: self.transport.clone(),
```

Comment:
ready occurs only once, so we could remove the `Clone` bound by making `transport` an `Option<T>` and using `take`.